File: src/api-scraper/typescript/reascript-USDocML.types.ts
export interface ReaScriptUSDocML {
  signatures: any[] | SignaturesClass
  requires: Requires
  tags: string[]
  target_document: TargetDocument
  source_document: SourceDocument
  description: Description
  chapters: Chapter[]
  return_values: Params
  params: Params
  functioncall: Functioncall
  title: string
}

export enum Chapter {
  APIHelperFunctions = "Api-Helper-Functions",
  AdditionalCFunctions = "Additional C++ Functions",
  AdditionalEELFunctions = "Additional EEL-Functions",
  AdditionalLuaFunctions = "Additional Lua-Functions",
  AdditionalPythonFunctions = "Additional Python-Functions",
  ArrangeviewManagement = "Arrangeview Management",
  AudioMIDIDevicemanagement = "AudioMidi Devicemanagement",
  AudioManagement = "Audio Management",
  AutomationManagement = "Automation Management",
  EnvelopeManagement = "Envelope Management",
  ExtendedStates = "Extended States",
  FXManagement = "FX Management",
  FileManagement = "File Management",
  JSPlugin = "JS_Plugin",
  JoystickManagement = "Joystick Management",
  MIDIManagement = "MIDI Management",
  MarkerManagement = "Marker Management",
  MediaitemManagement = "Mediaitem Management",
  MetaDataManagement = "MetaData Management",
  Miscellaneous = "Miscellaneous",
  Pitchshifting = "Pitchshifting",
  ProjectManagement = "Project Management",
  ReaPack = "ReaPack",
  ThemeManagement = "Theme management",
  TrackManagement = "Track Management",
  TransportManagement = "Transport Management",
  UserInterface = "User Interface",
}

export interface Description {
  language: string
  prog_lang: ProgLang
  description: string
  markup_type: MarkupType
}

export enum MarkupType {
  Markdown = "markdown",
  Plaintext = "plaintext",
}

export enum ProgLang {
  Eel = "eel",
  Empty = "*",
  ProgLang = "",
}

export interface Functioncall {
  lua?: string
  python?: string
  eel?: string
  cpp?: string
}

export interface Params {
  param_count?: number
  prog_lang?: ProgLang
  entries: Entry[]
  markup_type?: MarkupType
}

export interface Entry {
  description: string
  identifier: string
}

export interface Requires {
  Lua?: string
  Reaper: string
  SWS?: string
  JS?: string
  ReaPack?: string
}

export interface SignaturesClass {
  eel?: Eel
  lua?: Lua
}

export interface Eel {
  parameters: EelParameter[]
  return_values: string[]
  method_name?: string
}

export interface EelParameter {
  type_name?: string
  identifier: string
  modifier?: string
}

export interface Lua {
  parameters: ReturnValueElement[]
  return_values: ReturnValueElement[]
  method_name?: string
}

export interface ReturnValueElement {
  type: string
  identifier?: string
}

export enum SourceDocument {
  ReaperApidocsUSDocML = "reaper-apidocs.USDocML",
  ReaperPluginH = "reaper_plugin.h",
}

export enum TargetDocument {
  ReaperAPIDocumentation = "Reaper_Api_Documentation",
}

File: src/api-scraper/lua/reascript-USDocML.types.lua
---@meta

---@class ReaScriptUSDocML
---@field chapters Chapter[]
---@field description Description
---@field functioncall Functioncall
---@field params Params
---@field requires Requires
---@field return_values Params
---@field signatures any[] | SignaturesClass
---@field source_document SourceDocument
---@field tags string[]
---@field target_document TargetDocument
---@field title string

---@enum Chapter
local Chapter = {
  APIHelperFunctions = "Api-Helper-Functions",
  AdditionalCFunctions = "Additional C++ Functions",
  AdditionalEELFunctions = "Additional EEL-Functions",
  AdditionalLuaFunctions = "Additional Lua-Functions",
  AdditionalPythonFunctions = "Additional Python-Functions",
  ArrangeviewManagement = "Arrangeview Management",
  AudioMIDIDevicemanagement = "AudioMidi Devicemanagement",
  AudioManagement = "Audio Management",
  AutomationManagement = "Automation Management",
  EnvelopeManagement = "Envelope Management",
  ExtendedStates = "Extended States",
  FXManagement = "FX Management",
  FileManagement = "File Management",
  JSPlugin = "JS_Plugin",
  JoystickManagement = "Joystick Management",
  MIDIManagement = "MIDI Management",
  MarkerManagement = "Marker Management",
  MediaitemManagement = "Mediaitem Management",
  MetaDataManagement = "MetaData Management",
  Miscellaneous = "Miscellaneous",
  Pitchshifting = "Pitchshifting",
  ProjectManagement = "Project Management",
  ReaPack = "ReaPack",
  ThemeManagement = "Theme management",
  TrackManagement = "Track Management",
  TransportManagement = "Transport Management",
  UserInterface = "User Interface",
}

---@class Description
---@field language string
---@field prog_lang ProgLang
---@field description string
---@field markup_type MarkupType

---@enum MarkupType
local MarkupType = {
  Markdown = "markdown",
  Plaintext = "plaintext",
}

---@enum ProgLang
local ProgLang = {
  Eel = "eel",
  Empty = "*",
  ProgLang = "",
}

---@class Functioncall
---@field  lua string | nil
---@field  python string | nil
---@field  eel string | nil
---@field  cpp string | nil

---@class Params
---@field  param_count number | nil
---@field  prog_lang ProgLang | nil
---@field  entries Entry[]
---@field  markup_type MarkupType | nil

---@class Entry
---@field  description string
---@field  identifier string


---@class Requires
---@field  Lua string | nil
---@field  Reaper string
---@field  SWS string | nil
---@field  JS string | nil
---@field  ReaPack string | nil

---@class SignaturesClass
---@field  eel Eel | nil
---@field  lua Lua | nil

---@class Eel
---@field  parameters EelParameter[]
---@field  return_values string[]
---@field  method_name string

---@class EelParameter
---@field  type_name string | nil
---@field  identifier string
---@field  modifier string | nil

---@class Lua
---@field  parameters ReturnValueElement[]
---@field  return_values ReturnValueElement[]
---@field  method_name string | nil

---@class ReturnValueElement
---@field type string
---@field identifier string | nil

---@enum SourceDocument
local SourceDocument = {
  ReaperApidocsUSDocML = "reaper-apidocs.USDocML",
  ReaperPluginH = "reaper_plugin.h",
}

---@enum TargetDocument
local TargetDocument = {
  ReaperAPIDocumentation = "Reaper_Api_Documentation",
}
